### ISSUE TITLE
Feature request: a hybrid of merge, evolve and applySpec

### ISSUE THREAD

**Issue by arizonatribe** [ramda/ramda#2495]
I really like using `evolve()` and `applySpec()` to map API payloads into a nice presentational format for use in front-end components. Especially for large JSON structures the amount of typing it has saved me is tremendous. When using these functions I often encounter the scenario where I need to create new props on an object or where I need to create/update a prop that depends on multiple sources (other props), but this falls outside the scope of `evolve()`. Using `applySpec()` allows me to create new props that derive from one or more original props (say for example, I need to create a `fullName` prop that concatenates the original `firstName` and `lastName` props), however a limitation with `applySpec()` is passing through the original props unaltered. For the projects I've been working on I created a helper method called `mergeSpec()` that covers those cases where I need to map over an object and create new props and merge them back into the original object:

```
const mergeSpec = curry(function mergeSpec(spec, value) {
  return converge(merge, [identity, applySpec(spec)])(value)
})
```

I would use it in cases like this:

```
const formatUser = mergeSpec({
  fullName: R.compose(R.join(' '), R.values, R.pick(['firstName', 'lastName'])),
  address: R.pipe(
    R.prop('address'),
    R.evolve({
      street: R.trim,
      city: compose(str => str.replace(/(?:^|\s)\S/g, R.toUpper), R.trim),
      state: R.toUpper,
      zip: compose(R.trim, R.when(R.is(Number), R.toString))
    })
  )
});

formatUser({
  firstName: 'Montgomery',
  lastName: 'Burns',
  address: {
    street: '1000 Mammon Lane, ',
    city: 'springfield',
    state: 'or',
    zip: 97403
  }
}); // => { firsName: 'Montgomery', lastName: 'Burns', fullName: 'Montgomery Burns', address: { street: '1000 Mammon Lane,', city: 'Springfield', state: 'OR', zip: '97403' } }
```

It's been useful for me personally, but I'm not sure if it adds any value to the Ramda library or to Ramda-Adjunct. I've only been using Ramda for a year, and my functional programming experience prior was mostly fluent APIs, so I wouldn't be surprised if I haven't thought through some common pitfalls or debugging challenges this approach might bring, so I'm also fishing for critiques or better approaches to mapping objects in a more elegant way.

Thanks.

**Comment by CrossEye** [comment]
I think it's definitely worth creating a PR for this.  I'm guessing that we'd want an implementation that doesn't use `converge`, but that's just details.

What I'd really like would be a version that would also allow you to delete properties as well.  (It might make sense in your example to remove `firstName` and `lastName` as well.)  But I can't immediately think of a a good declarative API to handle this.  If we could find such a design, we might have a function that says, in a simple declarative fashion, "*take data that looks like* this *and change it to look like* that."  I believe that would be really powerful.

But even if we can't find that, I find this function to be persuasive.  It looks like a good fit.

**Comment by arizonatribe** [comment]
Okay, great! Here's a first pass at it. I'm not sure if its best to go this route or to import and use `applySpec` directly?
#2496

**Comment by CrossEye** [comment]
Reviewed with two minor corrections requested.

The point about `converge` is that it's an occasionally useful tool for where `lift` won't work, but all it does is to make it easier to write point-free code.  Our implementations are meant to be reasonably performant, and we prefer ugly imperative loops to nicely factored point-free code unless the latter can compete on performance.  `converge(multiply, [add, subtract])` is unlikely to compete with `function(a, b) {return (a + b) * (a - b);}`

> I did re-create much of the code from `applySpec`, but I'm not sure if it would have been better to import and use it directly.

If it makes the code much cleaner to do so, we'd certainly consider it.  But again, performance is always at least a minor concern.

**Comment by pkoch** [comment]
Would using `undefined` as a sentinel value to request a deletion be too bad?

**Comment by arizonatribe** [comment]
I'm not sure what happened with the underlying PR for this. I made the requested changes right away, but I assume there was just a lot of other fish to fry or it got forgotten. I ended up creating [another library called shapey](https://github.com/arizonatribe/shapey) to enrich the whole "spec mapping" functionality I was looking for. For that library I went with a reserved prop (called `shapeyMode`) that drives the removal, merge, etc. Maybe what you mentioned might work here; I dunno.

Ultimately though I found it was best to _combine_ the functionality of `evolve` and `applySpec` by making it so that the functions you create in an evolve spect have the signature of `(val, index, obj) -> *` rather than `val -> *`. It's more consistent with the way native JS `Array.prototype` methods work, and it kills to birds with one stone. I implemented that function into that library as [mapSpec](https://github.com/arizonatribe/vanillas/blob/master/src/mapSpec.js). Feel free to steal it if it's ever of interest; I was certainly influenced heavily by this library.

**Comment by CrossEye** [comment]
I'm not sure what happened either, except that this is a minor side project for me, and although there are many people who still contribute, @Bradcomp and I are the only ones actively involved right now with keeping this going.  If we both are distracted at the same time, things can get lost.  Sorry.

This does look like a useful function, and we may still take it.

Thank you again for your PR.

**Comment by arizonatribe** [comment]
no problem; I appreciate all the work!

**Comment by phdog** [comment]
`compose(chain(mergeDeepLeft), applySpec)`
this guy will applySpec and pass all rest props through

**Comment by arizonatribe** [comment]
That not quite the same thing, but still interesting, thanks.

---

I gave up on getting this merged into Ramda. I don't know why it got lost, but frankly I haven't used Ramda in a few years anyway. I switched over to something that's a lot faster (even faster than Lodash), called [vanillas](https://www.npmjs.com/package/vanillas).

I mentioned earlier this got added into an open-source library which specifically deals with reshaping objects using these same kinds of "specs". It uses `vanillas` under-the-hood rather than `ramda` so it's more performant. Check it out if you're interested
